sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+          + "(keepalive) called with null keepalive");
+      return builder().setScrollKeepalive(scrollKeepalive).build();
+    }
+
+    @Override
+    public PCollection<String> apply(PBegin input) {
+      return input.apply(org.apache.beam.sdk.io.Read.from(
+          new BoundedElasticsearchSource(this, null)
+      ));
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      checkState(getConnectionConfiguration() != null, "ElasticsearchIO.read() requires a "
+          + "connection configuration to be set via withConnectionConfiguration(configuration)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder.addIfNotNull(DisplayData.item("query", getQuery()));
+      getConnectionConfiguration().populateDisplayData(builder);
+    }
+
+  }
+
+  /**
+   * A {@link BoundedSource} reading from Elasticsearch.
+   */
+  protected static class BoundedElasticsearchSource extends BoundedSource<String> {
+
+    private ElasticsearchIO.Read spec;
+    @Nullable
+    private final String shardPreference;
+
+    protected BoundedElasticsearchSource(Read spec, String shardPreference) {
+      this.spec = spec;
+      this.shardPreference = shardPreference;
+    }
+
+    @Override
+    public List<? extends BoundedSource<String>> splitIntoBundles(long desiredBundleSizeBytes,
+                                                                  PipelineOptions options)
+        throws Exception {
+      List<BoundedElasticsearchSource> sources = new ArrayList<>();
+
+      // we split per shard
+      // unfortunately, Elasticsearch 2.x doesn't provide a way to do parallel reads on a single
+      // shard
+      JsonObject statsJson = getStats(true);
+      JsonObject shardsJson = statsJson.getAsJsonObject("indices")
+          .getAsJsonObject(spec.getConnectionConfiguration().getIndex())
+          .getAsJsonObject("shards");
+      Set<Map.Entry<String, JsonElement>> shards = shardsJson.entrySet();
+      for (Map.Entry<String, JsonElement> shardJson : shards) {
+        String shardId = shardJson.getKey();
+        JsonArray value = (JsonArray) shardJson.getValue();
+        boolean isPrimaryShard =
+            value.get(0).getAsJsonObject().getAsJsonObject("routing").getAsJsonPrimitive(
+                "primary").getAsBoolean();
+        if (isPrimaryShard) {
+          sources.add(new BoundedElasticsearchSource(spec, shardId));
+        }
+      }
+      return sources;
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws IOException {
+      // as Elasticsearch 2.x doesn't not support any way to do parallel read inside a shard
+      // the estimated size bytes is not really used in the split into bundles.
+      // However, we implement this method anyway as the runners can use it.
+      // NB: Elasticsearch 5.x now provides the slice API.
```

Comment:
The slide API is a new API coming into Elasticsearch 5. It will allow to not use shard anymore and use a Elasticsearch API really focusing on split read.